### DISCUSSION TITLE
Linked account: paged listing of Identity Providers

### DISCUSSION THREAD

**Discussion by laskasn** [keycloak/keycloak#8609] [Ideas]
### Motivation:

On keycloak’s account management console, if a realm has a lot of identity providers, the user is presented with a huge list which he has to scroll in order to find the desired identity provider to link his account with, since there is no search/filtering in the UI. Moreover, a lot of information is fetched, making the webpage quite resource demanding in terms of memory and network bandwidth.

### Design proposal:

We propose to perform minimal changes on the current design and add a paging and filtering on both {Linked Login Providers, Unlinked Login Providers} lists.

Our enhancement is done by making as few as possible changes on both the react and java sides.
So, with the following minimal changes: 

- on the backend side (java), we enabled LinkedAccountRepresentation items to be fetched in pages in LinkedAccountsResource.java
- on the frontend side (react/patternfly), we added a patternfly Pagination element to drive the requests of the pages from the backend and also added a filter, to filter out (server-side) the response

The loading times are hugely improved:

We have ~4300 IdPs in our keycloak and we measured the loading times and the idp-related data fetching.

  | Before | After
-- | -- | --
Data needed to fetch from /auth/realms/master/account/linked-accounts | 1,1 mb | 3,5 kb
Total time to load webpage | 7,49 sec | 0,834 sec

The pull request can be found **[here](https://github.com/keycloak/keycloak/pull/8022)**

We also attach two screenshots to showcase the UI changes:

#### Before:

![before](https://user-images.githubusercontent.com/11722418/137909446-048aa2c6-ec0f-4950-99fb-9efac2a25540.png)

#### After:

![after](https://user-images.githubusercontent.com/11722418/137909509-4412a00e-7dad-4bfd-a206-14e6dab758c0.png)

**Comment by stianst** [comment]
Looks like a nice addition for sure. I wonder if the pagination should be hidden if there's not many identity providers. In most cases there will never be more than a handful, and it would be strange to show pagination in those cases.

@ssilvert FIY

**Reply by laskasn** [reply]
Indeed, you 're right, in most cases there will be only a few idps on each realm and pagination would be strange.
You see, in all of our cases, we have a large number of IdPs (we also have a modified login theme to be able to serve the WAYF).
Anyway, that's something quite simple to change in the UI, i'll change it.

**Reply by laskasn** [reply]
I updated the code. Now it only shows paging if the actual items are more than 10 in any of the two lists (linked idps, unlinked idps).
So, as you see in the screenshots, in the first case, there's only one linked and only one unlinked, so, there are no paging bars shown.
In the second case, the unlinked idps are more than 10, thus, we utilize the existing paging.
![nopaging](https://user-images.githubusercontent.com/11722418/138237013-40bfb6aa-5051-4b5f-8f61-23035a580471.png)

![unlinked_paging](https://user-images.githubusercontent.com/11722418/138237133-47a3af33-7759-48bb-81ef-63c894e9d3b4.png)

**Reply by stianst** [reply]
Thanks, so the proposal is good to me, and we'll get someone to review the PR as quickly as we can.

**Reply by laskasn** [reply]
Thanks @stianst
We also have a similar (actually even better) enhancement for the admin console for keycloak setups with a significant number of IdPs per realm.
As you will see, it offers paging in idps listing, following exactly the UI design of the clients paging. Moreover, if you have a few idps (i.e. 10-15), no paging is shown (exactly as in the clients listing page in admin UI).  
It is a huge improvement (+scalable) if you have a quite a few Identity providers as we have in our environments. 
I think it is a very nice contribution to the community :)    
We have created a [discussion](https://github.com/keycloak/keycloak/discussions/8608) and a [pull request](https://github.com/keycloak/keycloak/pull/7960), please take a look whenever you have some time!

**Comment by laskasn** [comment]
@stianst It seems that we left this in an "accepted but not yet merged [PR](https://github.com/keycloak/keycloak/pull/8022)" state  :smile: 
I believe it's quite a nice addition to leave it out.
